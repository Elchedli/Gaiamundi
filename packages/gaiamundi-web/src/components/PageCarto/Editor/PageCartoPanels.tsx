import { Tab } from '@headlessui/react';
import classNames from 'classnames';
import { FC } from 'react';
import { PageCartoPanelData } from './PageCartoDataPanel/PageCartoDataPanel';
import { PageCartoIndicatorPanel } from './PageCartoIndicatorPanel/PageCartoIndicatorPanel';

export const PageCartoPanels: FC = () => {
  const tabs = [
    {
      id: 1,
      title: 'Hypertexte',
      content: <div className="text-center">en cours de développement</div>,
    },
    {
      id: 2,
      title: 'Données',
      content: <PageCartoPanelData />,
    },
    {
      id: 3,
      title: 'Indicateurs',
      content: <PageCartoIndicatorPanel />,
    },
  ];

  return (
    <div className="h-full w-full p-2" data-testid="pagecarto-panels">
      <Tab.Group>
        <Tab.List className="flex bg-blue-600 rounded-t-lg p-1 pb-0">
          {tabs.map((tab) => (
            <Tab
              key={tab.id}
              className={({ selected }) =>
                classNames(
                  'rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700',
                  'ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1',
                  selected
                    ? 'bg-white shadow'
                    : 'text-blue-100 hover:bg-white/[0.12] hover:text-white'
                )
              }
            >
              {tab.title}
            </Tab>
          ))}
        </Tab.List>
        <Tab.Panels className="h-full rounded-b-lg border border-blue-700">
          {tabs.map((tab) => (
            <Tab.Panel
              key={tab.id}
              className="px-2 pb-5 overflow-y-scroll"
              style={{ maxHeight: 'calc(100vh - 64px * 2)' }}
            >
              {tab.content}
            </Tab.Panel>
          ))}
        </Tab.Panels>
      </Tab.Group>
    </div>
  );
};
