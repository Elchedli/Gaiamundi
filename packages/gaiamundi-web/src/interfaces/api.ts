export interface ApiError {
  status?: number;
  name?: string;
  message?: string;
  description?: string;
  details?: {
    errors?: Array<{
      message: string;
      name: string;
      path: string[];
    }>;
  };
}

export interface ApiErrorResponse {
  data?: any;
  error?: ApiError;
}

export type ApiResponse<T> = T;

export type ApiPaginationMeta = {
  //
  start?: number;
  limit?: number;
  // OR
  page?: number;
  pageSize?: number;
  pageCount?: number;
  // Always
  total: number;
};

export type ApiData<T> = T & {
  id: number;
  created_at: string;
  updated_at: string;
};

export interface ApiDocument<T> {
  data: ApiData<T>;
  meta?: {
    availableLocales?: string[];
  };
}

export interface ApiCollection<T> {
  data: Array<ApiData<T>>;
  meta: {
    pagination: ApiPaginationMeta;
  };
}
