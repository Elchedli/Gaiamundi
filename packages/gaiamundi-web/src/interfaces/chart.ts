import {
  AreaChartIcon,
  BarChartIcon,
  BubbleChartIcon,
  ColumnChartIcon,
  LineChartIcon,
  LineColumnChartIcon,
  PieChartIcon,
  ScatterChartIcon,
} from 'components/Icons/Charts';
import { AnimationEasing, AnimationOptions } from 'eazychart-core';
import {
  AreaChart,
  AreaChartProps,
  BarChart,
  BarChartProps,
  BubbleChart,
  BubbleChartProps,
  ColumnChart,
  LineChart,
  LineChartProps,
  LineColumnChart,
  PieChart,
  PieChartProps,
  ScatterChart,
  ScatterChartProps,
} from 'eazychart-react';

type ChartConfigProps<T> = Omit<
  T,
  'scopedSlots' | 'animationOptions' | 'rawData'
> & {
  animationOptions: Omit<AnimationOptions, 'easing'> & {
    easing: AnimationEasing;
  };
};

export type ChartType =
  | 'column'
  | 'bar'
  | 'pie'
  | 'line'
  | 'scatter'
  | 'bubble'
  | 'area'
  | 'lineColumn';

export type ChartTypes = {
  [key in ChartType]: {
    Icon: React.FC;
    ChartComponent: React.FC;
  };
};

export type AllChartConfigProps = ChartConfigProps<BarChartProps> &
  ChartConfigProps<PieChartProps> &
  ChartConfigProps<LineChartProps> &
  ChartConfigProps<ScatterChartProps> &
  ChartConfigProps<BubbleChartProps> &
  ChartConfigProps<AreaChartProps>;

export type Chart = {
  name: string;
  type: ChartType;
  props: AllChartConfigProps;
};

export const CHART_TYPES: ChartTypes = {
  column: {
    Icon: ColumnChartIcon,
    ChartComponent: ColumnChart,
  },
  bar: {
    Icon: BarChartIcon,
    ChartComponent: BarChart,
  },
  pie: {
    Icon: PieChartIcon,
    ChartComponent: PieChart,
  },
  line: {
    Icon: LineChartIcon,
    ChartComponent: LineChart,
  },
  scatter: {
    Icon: ScatterChartIcon,
    ChartComponent: ScatterChart,
  },
  bubble: {
    Icon: BubbleChartIcon,
    ChartComponent: BubbleChart,
  },
  area: {
    Icon: AreaChartIcon,
    ChartComponent: AreaChart,
  },
  lineColumn: {
    Icon: LineColumnChartIcon,
    ChartComponent: LineColumnChart,
  },
};

export const getChartIconByType = (type: keyof typeof CHART_TYPES) => {
  return type in CHART_TYPES ? CHART_TYPES[type].Icon : null;
};

export type RawDatumType =
  | 'string'
  | 'number'
  | 'bigint'
  | 'boolean'
  | 'symbol'
  | 'undefined'
  | 'object'
  | 'function';

export type RawDatum = Record<string, any>;
