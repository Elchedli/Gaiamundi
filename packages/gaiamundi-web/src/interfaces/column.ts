export interface Column {
  name: string;
  sample: number;
  source: string;
  validity: string;
  isGeoCode: boolean;
}

export interface DatasetColumn extends Column {
  dataset: string;
}
