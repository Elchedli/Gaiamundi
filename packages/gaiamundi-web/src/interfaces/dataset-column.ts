import { ApiDocument } from './api';
import { Dataset } from './dataset';
import { User } from './user';

export interface DatasetColumnBase {
  name: string;
  source: string;
  validity: boolean;
}

export interface DatasetColumnStub extends DatasetColumnBase {
  owner: number;
  dataset: number;
}

export interface DatasetColumn extends DatasetColumnBase {
  owner: ApiDocument<User>;
  dataset: ApiDocument<Dataset>;
}
