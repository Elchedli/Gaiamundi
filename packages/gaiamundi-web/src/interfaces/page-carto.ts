import { ApiData, ApiDocument } from './api';
import { GeoMapAttributes } from './geo-map';
import { User } from './user';

export interface Tags {
  id: number;
  name: string;
  type: string[];
}

export interface PageCartoAttributes {
  id?: string;
  name: string;
  owner?: ApiDocument<User>;
  map: GeoMapAttributes;
  html: HTMLElement;
  tags: Tags;
  cover?: MediaImage;
}

export type PageCarto = ApiData<PageCartoAttributes>;
