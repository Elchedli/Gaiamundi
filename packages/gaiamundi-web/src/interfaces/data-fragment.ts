import { ApiCollection, ApiDocument } from './api';
import { Column } from './column';
import { Dataset } from './dataset';

export interface DataFragmentStub {
  columns: number[];
  dataset: number;
  page_carto: number;
}

export interface DataFragment {
  columns: ApiCollection<Column>;
  dataset: ApiDocument<Dataset>;
}
