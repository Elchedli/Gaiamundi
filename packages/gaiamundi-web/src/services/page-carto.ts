import { PageCarto, PageCartoStub } from 'interfaces/page-carto';
import { ContentType, strapi } from './strapi';

export const createPageCarto = async (data: PageCartoStub) => {
  return await strapi.create<PageCartoStub>(ContentType.PAGE_CARTOS, data);
};

export const getLatestPageCartos = async (page: number, pageSize: number) => {
  return await strapi.get<PageCarto>(ContentType.PAGE_CARTOS, {
    populate: '*',
    sort: 'createdAt:desc',
    pagination: {
      page,
      pageSize,
    },
  });
};

export const getPageCartoById = async (id: number) => {
  return await strapi.getById<PageCarto>(ContentType.PAGE_CARTOS, id, {
    populate: {
      map: {
        populate: {
          geoJSON: true,
        },
      },
      owner: true,
      columns: {
        populate: {
          dataset: true,
        },
      },
    },
  });
};

export const uploadCsv = async (file: File) => {
  return await strapi.uploadFile(file, 'api::page-carto.page-carto');
};
