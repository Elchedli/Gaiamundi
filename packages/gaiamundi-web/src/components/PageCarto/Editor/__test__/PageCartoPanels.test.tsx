import { fireEvent, render } from '@testing-library/react';
import { ApiData } from 'interfaces/api';
import { PageCarto } from 'interfaces/page-carto';
import {
  mockCoverApiDocument,
  mockDataFragmentsApiCollection,
  mockGeoMapApiDocument,
  mockOwnerApiDocument,
  mockTagsApiCollection,
} from 'utils/mocks/data';
import { PageCartoPanels } from '../PageCartoPanels';
describe('PageCartoPanels', () => {
  const mockPageCartoData: ApiData<PageCarto> = {
    id: 1,
    attributes: {
      name: 'Page Carto 1',
      owner: mockOwnerApiDocument,
      tags: mockTagsApiCollection,
      map: mockGeoMapApiDocument,
      cover: mockCoverApiDocument,
      html: '<p>Test HTML</p>',
      data_fragments: mockDataFragmentsApiCollection,
      created_at: '2022-02-07T00:00:00.000Z',
      updated_at: '2022-02-07T00:00:00.000Z',
    },
  };

  it('renders the tabs and their content', async () => {
    const { getByText, container } = render(
      <PageCartoPanels pageCarto={mockPageCartoData} />
    );

    const tab1 = getByText('Hypertexte');
    expect(tab1).toBeInTheDocument();

    const tab2 = getByText('Donn√©es');
    expect(tab2).toBeInTheDocument();

    const tab3 = getByText('Indicateurs');
    expect(tab3).toBeInTheDocument();

    fireEvent.click(tab2);

    expect(getByText('secondColumn')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
