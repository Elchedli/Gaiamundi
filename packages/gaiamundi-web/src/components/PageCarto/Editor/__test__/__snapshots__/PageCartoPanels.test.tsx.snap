// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageCartoPanels renders the tabs and their content 1`] = `
<div>
  <div
    class="h-full w-full p-2"
  >
    <div
      aria-orientation="horizontal"
      class="flex bg-blue-600 rounded-t-lg p-1 pb-0"
      role="tablist"
    >
      <button
        aria-controls="headlessui-tabs-panel-:r3:"
        aria-selected="false"
        class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 text-blue-100 hover:bg-white/[0.12] hover:text-white"
        data-headlessui-state=""
        id="headlessui-tabs-tab-:r0:"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Hypertexte
      </button>
      <button
        aria-controls="headlessui-tabs-panel-:r4:"
        aria-selected="true"
        class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 bg-white shadow"
        data-headlessui-state="selected"
        id="headlessui-tabs-tab-:r1:"
        role="tab"
        tabindex="0"
        type="button"
      >
        Données
      </button>
      <button
        aria-controls="headlessui-tabs-panel-:r5:"
        aria-selected="false"
        class="rounded-t-lg p-2 text-sm font-medium leading-5 text-blue-700 ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 text-blue-100 hover:bg-white/[0.12] hover:text-white"
        data-headlessui-state=""
        id="headlessui-tabs-tab-:r2:"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Indicateurs
      </button>
    </div>
    <div
      class="h-full rounded-b-lg border border-blue-700"
    >
      <span
        aria-labelledby="headlessui-tabs-tab-:r0:"
        id="headlessui-tabs-panel-:r3:"
        role="tabpanel"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        tabindex="-1"
      />
      <div
        aria-labelledby="headlessui-tabs-tab-:r1:"
        class="px-2 pb-5 overflow-y-scroll"
        data-headlessui-state="selected"
        id="headlessui-tabs-panel-:r4:"
        role="tabpanel"
        style="max-height: calc(100vh - 64px * 2);"
        tabindex="0"
      >
        <div>
          <div
            class="mt-5 text-right"
          >
            <button
              class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600"
              data-testid="button-element"
              type="button"
            >
              <span
                class="flex items-center text-sm px-4 py-2 rounded-md !px-2"
              >
                <svg
                  aria-hidden="true"
                  class="w-5 h-5"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12 3.75a.75.75 0 01.75.75v6.75h6.75a.75.75 0 010 1.5h-6.75v6.75a.75.75 0 01-1.5 0v-6.75H4.5a.75.75 0 010-1.5h6.75V4.5a.75.75 0 01.75-.75z"
                    fill-rule="evenodd"
                  />
                </svg>
                Importer un jeu de données
              </span>
            </button>
          </div>
          <div
            class="my-2"
          >
            <div
              aria-colcount="4"
              aria-rowcount="4"
              class="rdg r104f42s7-0-0-beta-22 border"
              dir="ltr"
              role="grid"
              style="grid-template-rows: 35px repeat(3, 35px); --rdg-header-row-height: 35px; --rdg-summary-row-height: 35px; --rdg-sign: 1; grid-template-columns: auto auto auto auto;"
            >
              <div
                aria-rowindex="1"
                class="rdg-header-row h197vzie7-0-0-beta-22"
                role="row"
                style="--rdg-grid-row-start: 1;"
              >
                <div
                  aria-colindex="1"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="columnheader"
                  style="grid-column-start: 1;"
                  tabindex="0"
                >
                  Colonne
                </div>
                <div
                  aria-colindex="2"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="columnheader"
                  style="grid-column-start: 2;"
                  tabindex="-1"
                >
                  Source
                </div>
                <div
                  aria-colindex="3"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="columnheader"
                  style="grid-column-start: 3;"
                  tabindex="-1"
                >
                  Validité
                </div>
                <div
                  aria-colindex="4"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="columnheader"
                  style="grid-column-start: 4;"
                  tabindex="-1"
                >
                  Jeu de données
                </div>
              </div>
              <div
                aria-rowindex="2"
                class="rdg-row r1otpg647-0-0-beta-22 rdg-row-even"
                role="row"
                style="--rdg-grid-row-start: 2; --rdg-row-height: 35px;"
              >
                <div
                  aria-colindex="1"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 1;"
                  tabindex="-1"
                >
                  secondColumn
                </div>
                <div
                  aria-colindex="2"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 2;"
                  tabindex="-1"
                >
                  source2
                </div>
                <div
                  aria-colindex="3"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 3;"
                  tabindex="-1"
                >
                  2022
                </div>
                <div
                  aria-colindex="4"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 4;"
                  tabindex="-1"
                >
                  firstDataset
                </div>
              </div>
              <div
                aria-rowindex="3"
                class="rdg-row r1otpg647-0-0-beta-22 rdg-row-odd"
                role="row"
                style="--rdg-grid-row-start: 3; --rdg-row-height: 35px;"
              >
                <div
                  aria-colindex="1"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 1;"
                  tabindex="-1"
                >
                  thirdColumn
                </div>
                <div
                  aria-colindex="2"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 2;"
                  tabindex="-1"
                >
                  source3
                </div>
                <div
                  aria-colindex="3"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 3;"
                  tabindex="-1"
                >
                  2023
                </div>
                <div
                  aria-colindex="4"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 4;"
                  tabindex="-1"
                >
                  firstDataset
                </div>
              </div>
              <div
                aria-rowindex="4"
                class="rdg-row r1otpg647-0-0-beta-22 rdg-row-even"
                role="row"
                style="--rdg-grid-row-start: 4; --rdg-row-height: 35px;"
              >
                <div
                  aria-colindex="1"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 1;"
                  tabindex="-1"
                >
                  fourthColumn
                </div>
                <div
                  aria-colindex="2"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 2;"
                  tabindex="-1"
                >
                  source4
                </div>
                <div
                  aria-colindex="3"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 3;"
                  tabindex="-1"
                >
                  2024
                </div>
                <div
                  aria-colindex="4"
                  aria-readonly="true"
                  aria-selected="false"
                  class="rdg-cell c1wupbe7-0-0-beta-22"
                  role="gridcell"
                  style="grid-column-start: 4;"
                  tabindex="-1"
                >
                  firstDataset
                </div>
              </div>
              <div
                class="m1l09lto7-0-0-beta-22"
                data-measuring-cell-key="name"
                style="grid-column-start: 1; min-width: 80px;"
              />
              <div
                class="m1l09lto7-0-0-beta-22"
                data-measuring-cell-key="source"
                style="grid-column-start: 2; min-width: 80px;"
              />
              <div
                class="m1l09lto7-0-0-beta-22"
                data-measuring-cell-key="validity"
                style="grid-column-start: 3; min-width: 80px;"
              />
              <div
                class="m1l09lto7-0-0-beta-22"
                data-measuring-cell-key="dataset"
                style="grid-column-start: 4; min-width: 80px;"
              />
            </div>
          </div>
        </div>
      </div>
      <span
        aria-labelledby="headlessui-tabs-tab-:r2:"
        id="headlessui-tabs-panel-:r5:"
        role="tabpanel"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
